src/libraries/System.Text.Json/src/System/Text/Json/Serialization/AsyncEnumerableStreamingDeserializer.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Threading;
+
+namespace System.Text.Json.Serialization
+{
+    internal sealed class AsyncEnumerableStreamingDeserializer<TValue> : IAsyncEnumerable<TValue>
```

Comment:
Is this type still needed?  It appears to be used in one place:
```C#
public static IAsyncEnumerable<TValue> DeserializeAsyncEnumerable<[DynamicallyAccessedMembers(MembersAccessedOnRead)] TValue>(
    Stream utf8Json,
    JsonSerializerOptions? options = null)
{
    if (utf8Json == null)
    {
        throw new ArgumentNullException(nameof(utf8Json));
    }

    return new AsyncEnumerableStreamingDeserializer<TValue>(utf8Json, options);
}
```
which can instead be:
```C#
public static IAsyncEnumerable<TValue> DeserializeAsyncEnumerable<[DynamicallyAccessedMembers(MembersAccessedOnRead)] TValue>(
    Stream utf8Json,
    JsonSerializerOptions? options = null)
{
    if (utf8Json == null)
    {
        throw new ArgumentNullException(nameof(utf8Json));
    }

    return Deserialize(utf8Json ?? throw new ArgumentNullException(nameof(utf8Json)), options);

    private async IAsyncEnumerator<TValue> Deserialize(Stream utf8Json, JsonSerializerOptions? options = null, [EnumeratorCancellation] CancellationToken cancellationToken)
    {
        ... // current body of your method
    }
}
```
That will be simpler (less ceremony) and more efficient in that the compiler will enable reusing the enumerable as the enumerator.

Reply:
(Also note the [EnumeratorCancellation] attribute, which enables a CancellationToken passed to GetAsyncEnumerator to be passed in as this argument.)

Reply:
Am I right to assume then that this code doesn't in fact hardcode `cancellationToken` to `default`, and instead compiler magic will make sure that the right token is passed at enumeration time? 🤯

Reply:
Yes.  There are two possible tokens here: one passed directly to IAsyncEnumerable-returning method, and one passed to its GetAsyncEnumerator.  The compiler generates logic akin to:
- If one is default, use the other.
- If they're the same, use it.
- If both are non-default, use CancellationTokenSource.CreateLinkedTokenSource to create a new token that combines them (and then Dispose on the enumerator appropriately disposes that CTS).

If you look at https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQGADAAQY4oDchJ5OAdAEoCuAdgC4CWAtjAwGEIPAA5cANjCgBlKQDcuYGAGcaRMhQYAVABawAhgBMubAOZraAZnJZSA0gG9CpF+WsYAHKQCSAQWUAnmxgAKJsLHxQ+sCSADwmHAB8pACyABQA2mERUvoc0AL6wTDi4nlcEGwAunZFSqXllVoQANYwbKRgdSVl3E2t7QCUzq5OBK4Tnd0NfWzNbWzaehAA7t4AZoXFMxVsTDAAjiwqHDCGaYNqk64UOKTABi1XrgC+hC9AA==, for example, you can see that generated logic in the GetAsyncEnumerator method:
```C#
            if (<>3__cancellationToken.Equals(default(CancellationToken)))
            {
                <M>d__.cancellationToken = cancellationToken;
            }
            else if (cancellationToken.Equals(<>3__cancellationToken) || cancellationToken.Equals(default(CancellationToken)))
            {
                <M>d__.cancellationToken = <>3__cancellationToken;
            }
            else
            {
                <>x__combinedTokens = CancellationTokenSource.CreateLinkedTokenSource(<>3__cancellationToken, cancellationToken);
                <M>d__.cancellationToken = <>x__combinedTokens.Token;
            }
```

Reply:
When I tried this out I was confused by the fact that calling the method explicitly still required that parameter, instead of it being implicit-like.